src/java/org/apache/cassandra/streaming/StreamSession.java

```diff
@@ -1027,9 +1022,30 @@ public void receive(IncomingStreamMessage message)
     public void progress(String filename, ProgressInfo.Direction direction, long bytes, long total)
     {
         ProgressInfo progress = new ProgressInfo(peer, index, filename, direction, bytes, total);
+        updateMetricsOnProgress(progress);
         streamResult.handleProgress(progress);
     }
 
+    private void updateMetricsOnProgress(ProgressInfo progress)
+    {
+        ProgressInfo.Direction direction = progress.direction;
+        long lastSeenBytesStreamedForProgress = lastSeenBytesStreamed.getOrDefault(progress, 0L);
+        long newBytesStreamed = progress.currentBytes - lastSeenBytesStreamedForProgress;
+        if (direction == ProgressInfo.Direction.OUT)
+        {
+            StreamingMetrics.totalOutgoingBytes.inc(newBytesStreamed);
+            metrics.outgoingBytes.inc(newBytesStreamed);
+        }
+
+        else if (direction == ProgressInfo.Direction.IN)
+        {
+            StreamingMetrics.totalIncomingBytes.inc(newBytesStreamed);
+            metrics.incomingBytes.inc(newBytesStreamed);
+        }
+
+        lastSeenBytesStreamed.put(progress, lastSeenBytesStreamedForProgress + newBytesStreamed);
```

Comment:
In practice we will be adding a new progress object for every file streamed in each direction by this `StreamSession` (see: [ProgressInfo::hashCode](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/streaming/ProgressInfo.java#L98)). The worst case is the number of total entries is equal to number of files on the local node * number of files on the remote node for each `StreamSession`. 

A simple optimization I can add here would be to remove the progress object from the map once we've completed streaming. In this case when `progress.currentBytes == progress.totalBytes`.  This would clean things up as each file completes and probably improve memory utilization in the average case. This wouldn't handle the worst case where all files complete streaming at the same time.

Reply:
_ The worst case is the number of total entries is equal to number of files on the local node * number of files on the remote node for each StreamSession._

Is it really _times_ (multiplication)? So thousand files on each side is 1 million progress objects? Or you meant _plus_ ?

Anyway I guess removing it would help indeed, we probably need to be cautions what to do when a streaming session fails in the middle and it is just garbage collected or something similar ...

Reply:
Good catch, yeah `+`.

Reply:
also, would it be enough to pass just `bytes` used upon ProgressInfo instantiation into `updateMetricsOnProgress` method as another argument? So you would not need to bypass that object at all, probably. I am just trying to find some workaround.

Reply:
I think you ultimately need just file name, direction and bytes.

Reply:
````
        
        Map<String, long[]> ins = new HashMap<>();
        Map<String, long[]> outs = new HashMap<>();

        // at the beginning we would put it in there based on direction        
        ins.put(progress.fileName, new long[] {0, progress.totalBytes});
````

So later we could check when to remove it from that map as we have access to `totalBytes` we previously put there.

Reply:
Yes, what we really care about to identify the uniqueness is the `(fileName, direction)` tuple, so two maps could work.

I'm not really sure about why you're looking to store the `progress.totalBytes` value in the `ins` and `outs` maps.  The call to `progress` _should_ always set the same value for a given `(fileName, direction)` pair.

Reply:
too hard to elaborate on that here and now, lets see what you come up with and then we will see

Reply:
I think that main reason for putting there totalBytes was to know when you are allowed to remove that entry from the map based on your `progress.currentBytes == progress.totalBytes` suggestion.

Reply:
I tried out removing from the map when `progress.currentBytes == progress.totalBytes` and found that this function is called in [CassandraCompressedStreamReader::read](https://github.com/apache/cassandra/blob/trunk/src/java/org/apache/cassandra/db/streaming/CassandraCompressedStreamReader.java#L96) multiple times (seemingly once per partition written as a file). What this results in is `progress` being called multiple times where `currentBytes == totalBytes`.  I'm guessing this is a behavior we'd like to keep instead of moving the call to progress outside of the `while` loop in `CassandraCompressedStreamReader::read`.

Given that, I'm going to spend some time thinking of another way to clean up this map when file streaming is done.